core/toolbox/category.js

```diff
@@ -416,3 +416,184 @@ Blockly.ToolboxCategory.prototype.parseColour_ = function(colourValue) {
+  if (isExpanded) {
+    this.subcategoriesDiv_.style.display = 'block';
+    this.openIcon_(this.iconSpan_);
+  } else {
+    this.subcategoriesDiv_.style.display = 'none';
+    this.closeIcon_(this.iconSpan_);
+  }
+  Blockly.utils.aria.setState(/** @type {!Element} */ (this.htmlDiv_),
+      Blockly.utils.aria.State.EXPANDED, isExpanded);
+
+  if (this.hasChildren()) {
+    for (var i = 0; i < this.contents_.length; i++) {
+      var child = this.contents_[i];
+      child.setVisible(isExpanded);
+    }
+  }
+
+  Blockly.svgResize(this.workspace_);
+};
+
+/**
+ * Adds appropriate classes to display an open icon.
+ * @param {?Element} iconDiv The div that holds the icon.
+ * @protected
+ */
+Blockly.ToolboxCategory.prototype.openIcon_ = function(iconDiv) {
+  if (!iconDiv) {
+    return;
+  }
+
+  Blockly.utils.dom.removeClass(iconDiv, this.classConfig_['closedIcon']);
+  Blockly.utils.dom.addClass(iconDiv, this.classConfig_['openIcon']);
+};
+
+/**
+ * Adds appropriate classes to display a closed icon.
+ * @param {?Element} iconDiv The div that holds the icon.
+ * @protected
+ */
+Blockly.ToolboxCategory.prototype.closeIcon_ = function(iconDiv) {
+  if (!iconDiv) {
+    return;
+  }
+  Blockly.utils.dom.removeClass(iconDiv, this.classConfig_['openIcon']);
+  Blockly.utils.dom.addClass(iconDiv, this.classConfig_['closedIcon']);
+};
+
+/**
+ * Whether or not this category has subcategories.
+ * @return {boolean} True if this category has subcategories, false otherwise.
+ * @public
+ */
+Blockly.ToolboxCategory.prototype.hasChildren = function() {
+  return this.hasChildren_;
+};
+
+/**
+ * Sets whether the category is visible or not. Categories are not visible if
+ * they are the child of a parent who has been collapsed.
+ * @param {boolean} isVisible True if category should be visible.
+ * @public
+ */
+Blockly.ToolboxCategory.prototype.setVisible = function(isVisible) {
+  // TODO: Add ability to hide the category when this is false.
+  //  This causes problems for nested categories.
+  this.isVisible_ = isVisible;
+};
+
+/**
+ * Whether the category is visible.
+ * @return {boolean} True if the category is visible, false otherwise.
+ * @public
+ */
+Blockly.ToolboxCategory.prototype.isVisible = function() {
+  return this.isVisible_;
+};
+
+/**
+ * @override
+ */
+Blockly.ToolboxCategory.prototype.isExpanded = function() {
+  return this.expanded_;
+};
+
+/**
+ * @override
+ */
+Blockly.ToolboxCategory.prototype.isSelectable = function() {
+  // TODO: Add && isSelectable_
+  return this.isVisible();
+};
+
+/**
+ * @override
+ */
+Blockly.ToolboxCategory.prototype.isCollapsible = function() {
+  return this.hasChildren();
+};
+
+/**
+ * @override
+ */
+Blockly.ToolboxCategory.prototype.onClick = function(_e) {
+  this.toggleExpanded();
+};
+
+/**
+ * @override
+ */
+Blockly.ToolboxCategory.prototype.setSelected = function(isSelected) {
+  if (isSelected) {
+    this.rowDiv_.style.backgroundColor = this.colour_ || '#57e';
```

Comment:
`#57e` should probably be a default color property instead.

Reply:
done.